### PR TITLE
Add support to ActiveRecord 7.0

### PR DESCRIPTION
**Pull request by gridanjbf**

### CODE REVIEW

File: spec/chrono_model/time_machine/manipulations_spec.rb [hunk 1]
```diff
@@ -25,7 +25,7 @@
     it do
       with_revert do
         subject.name = 'another modified bar history'
-        subject.save
+        subject.save!
```
Comment (tagliala) [inline spec/chrono_model/time_machine/manipulations_spec.rb:28]: This can be backported to master, correct?

Comment (gridanjbf) [reply spec/chrono_model/time_machine/manipulations_spec.rb:28]: correct. The spec states we want to test save! but then save! is called. I guess it was a long term typo.

Comment (lleirborras) [reply spec/chrono_model/time_machine/manipulations_spec.rb:28]: yes, it looks like @vjt was wrong

Comment (vjt) [reply spec/chrono_model/time_machine/manipulations_spec.rb:28]: it has been a long-term exercise for you padawans. good job in spotting it. strong the force in you is

File: .github/workflows/ruby.yml [hunk 2]
```diff
@@ -27,18 +27,34 @@ jobs:
...
+        gemfile: [ rails_5.0, rails_5.1, rails_5.2, rails_6.0, rails_6.1, rails_7.0.0.alpha2, rails_7.0.0.edge ]
         experimental: [false]
 
         exclude:
           - ruby-version: '2.3'
             gemfile: rails_6.0
           - ruby-version: '2.3'
             gemfile: rails_6.1
+          - ruby-version: '2.3'
+            gemfile: rails_7.0.0.alpha2
```
Comment (tagliala) [inline .github/workflows/ruby.yml:39]: Please use `7.0` for uniformity (it is ok for now to omit alpha)

File: .github/workflows/ruby.yml [hunk 3]
```diff
@@ -27,18 +27,34 @@ jobs:
...
+          - ruby-version: '2.3'
+            gemfile: rails_7.0.0.edge
```
Comment (tagliala) [inline .github/workflows/ruby.yml:41]: `rails_edge`

File: .github/workflows/ruby.yml [hunk 4]
```diff
@@ -27,18 +27,34 @@ jobs:
...
+          - ruby-version: '2.4'
+            gemfile: rails_7.0.0.edge
+          - ruby-version: '2.5'
+            gemfile: rails_7.0.0.alpha2
+          - ruby-version: '2.5'
+            gemfile: rails_7.0.0.edge
+          - ruby-version: '2.6'
+            gemfile: rails_7.0.0.alpha2
+          - ruby-version: '2.6'
+            gemfile: rails_7.0.0.edge
```
Comment (tagliala) [inline .github/workflows/ruby.yml:57]: `rails_7.0` an `rails_edge` for uniformity

File: spec/aruba/fixtures/railsapp/config/initializers/assets.rb [hunk 5]
```diff
@@ -0,0 +1,3 @@
+## this file overrides the default initialization of assets.rb
+
+ Rails.application.config.assets.version = "1.0" if Rails.version < '7.0'
```
Comment (tagliala) [inline spec/aruba/fixtures/railsapp/config/initializers/assets.rb:3]: missing new line at the end of the file

File: spec/aruba/fixtures/railsapp/config/environments/development.rb [hunk 6]
```diff
@@ -24,10 +24,13 @@
...
   # This option may cause significant delays in view rendering with a large
   # number of complex assets.
-  config.assets.debug = true
+
+  if Rails.version < '7.0'
+    config.assets.debug = true
 
   # Suppress logger output for asset requests.
-  config.assets.quiet = true
+   config.assets.quiet = true
```
Comment (tagliala) [inline spec/aruba/fixtures/railsapp/config/environments/development.rb:32]: wrong indentation

File: lib/chrono_model/patches/relation.rb [hunk 7]
```diff
@@ -4,6 +4,26 @@ module Patches
     module Relation
       include ChronoModel::Patches::AsOfTimeHolder
 
+      ## remove preload for now since is just an optimization
+      def preload_associations(records) # :nodoc:
+        return super if ActiveRecord::VERSION::STRING < '7.0'
+        preload = preload_values
```
Comment (tagliala) [inline lib/chrono_model/patches/relation.rb:10]: please add a new empty line here to separate the guard clause

File: Rakefile [hunk 8]
```diff
@@ -16,7 +17,11 @@ namespace :testapp do
     FileUtils.mkdir_p('tmp/aruba')
     Dir.chdir('tmp') do
       FileUtils.rm_rf('railsapp')
+      if Rails.version < '7.0'
       sh 'rails new railsapp --skip-bundle --skip-webpack-install --skip-git'
```
Comment (tagliala) [inline Rakefile:21]: please fix indentation

File: lib/chrono_model/patches/relation.rb [hunk 9]
```diff
@@ -4,6 +4,26 @@ module Patches
     module Relation
       include ChronoModel::Patches::AsOfTimeHolder
 
+      ## remove preload for now since is just an optimization
```
Comment (tagliala) [inline lib/chrono_model/patches/relation.rb:7]: is this comment still relevant?

File: spec/aruba/fixtures/railsapp/config/environments/development.rb [hunk 10]
```diff
@@ -29,7 +29,7 @@
     config.assets.debug = true
 
   # Suppress logger output for asset requests.
```
Comment (tagliala) [inline spec/aruba/fixtures/railsapp/config/environments/development.rb:31]: also this comment

File: spec/aruba/fixtures/railsapp/config/initializers/assets.rb [hunk 11]
```diff
@@ -1,3 +1,4 @@
 ## this file overrides the default initialization of assets.rb
 
- Rails.application.config.assets.version = "1.0" if Rails.version < '7.0'
\ No newline at end of file
+ Rails.application.config.assets.version = "1.0" if Rails.version < '7.0'
+
```
Comment (tagliala) [inline spec/aruba/fixtures/railsapp/config/initializers/assets.rb:4]: there is still a missing new line at the end of the file, this may be some editor configuration

image

the file should not end with the warning

File: lib/active_record/tasks/chronomodel_database_tasks.rb [hunk 12]
```diff
@@ -22,7 +22,8 @@ def structure_dump(*arguments)
       end
 
       def data_dump(target)
-        set_psql_env
+        set_psql_env if ActiveRecord::VERSION::STRING < '7.0'
+        psql_env if ActiveRecord::VERSION::STRING >= '7.0'
```
Comment (tagliala) [inline lib/active_record/tasks/chronomodel_database_tasks.rb:26]: Can we go for `if..else` here?

File: lib/active_record/tasks/chronomodel_database_tasks.rb [hunk 13]
```diff
@@ -31,7 +32,8 @@ def data_dump(target)
       end
 
       def data_load(source)
-        set_psql_env
+        set_psql_env if ActiveRecord::VERSION::STRING < '7.0'
+        psql_env if ActiveRecord::VERSION::STRING >= '7.0'
```
Comment (tagliala) [inline lib/active_record/tasks/chronomodel_database_tasks.rb:36]: same as above

File: lib/active_record/tasks/chronomodel_database_tasks.rb [hunk 14]
```diff
@@ -43,6 +41,14 @@ def data_load(source)
 
       private
 
+      def set_psql_env
+        if ActiveRecord::VERSION::STRING < '7.0'
+          super
+        else
+          psql_env
+        end
```
Comment (tagliala) [inline lib/active_record/tasks/chronomodel_database_tasks.rb:49]: much better, thanks

Comment (gridanjbf) [reply lib/active_record/tasks/chronomodel_database_tasks.rb:49]: thanks :)

File: spec/aruba/fixtures/railsapp/config/environments/development.rb [hunk 15]
```diff
@@ -24,10 +24,10 @@
   # Debug mode disables concatenation and preprocessing of assets.
   # This option may cause significant delays in view rendering with a large
   # number of complex assets.
-  config.assets.debug = true
+  config.assets.debug = true if Rails.version < '7.0'
 
   # Suppress logger output for asset requests.
-  config.assets.quiet = true
+  config.assets.quiet = true if Rails.version < '7.0'
```
Comment (tagliala) [inline spec/aruba/fixtures/railsapp/config/environments/development.rb:30]: two `if`s to reduce the number of changes

**Comment by tagliala** [comment]
@gridanjbf I've rebased from master and cleaned up the commits

Meanwhile, something changed in ruby edge and 

```
Don't know how to build task 'db:structure:load' (See the list of available tasks with `rake --tasks`)
       Did you mean?  db:test:load

rspec ./spec/aruba/rake_task_spec.rb:11 # rake tasks bundle exec rake -T is expected to have output: /db:structure:load/
rspec ./spec/aruba/rake_task_spec.rb:62 # rake tasks db:structure:load given a file db/structure.sql with default username and password is expected to be successfully executed
rspec ./spec/aruba/rake_task_spec.rb:68 # rake tasks db:structure:load given a file db/structure.sql without a specified username and password is expected to be successfully executed
```

The old branch is available [here](https://github.com/ifad/chronomodel/tree/tmp/feature%2Far-70) and specs are failing there too

**Comment by tagliala** [comment]
Ref: rails/rails@cf269d51fbc4758cb6ae0110c6c151546c0889d9

File: Rakefile [hunk 16]
```diff
@@ -1,4 +1,5 @@
 require "bundler/gem_tasks"
+require "rails"
```
Comment (tagliala) [inline Rakefile:2]: Let's use `require 'rails/version'`

```
$ time ruby -e 'require "rails"'

real	0m0.454s
user	0m0.337s
sys	0m0.103s

$ time ruby -e 'require "rails/version"'

real	0m0.125s
user	0m0.085s
sys	0m0.031s
```

File: spec/support/aruba.rb [hunk 17]
```diff
@@ -1,9 +1,32 @@
 require 'aruba/rspec'
 require 'arel'
+require 'rake'
+require 'rails'
 
 module ChronoTest
 
   module Aruba
+
+    def load_schema_task(as_regexp = false)
```
Comment (tagliala) [inline spec/support/aruba.rb:10]: let's use keyword arguments, `load_schema_task(as_regexp: false)` to improve this method call: `load_schema_task(true)` is not clear to understand

File: spec/support/aruba.rb [hunk 18]
```diff
@@ -1,9 +1,32 @@
...
+      str =
+      if Rails.version < '7.0'
+        "db:structure:load"
+      else
+        "db:schema:load"
+      end
+      return as_regexp ? Regexp.new(str) : str
```
Comment (tagliala) [inline spec/support/aruba.rb:17]: useless return

File: spec/support/aruba.rb [hunk 19]
```diff
@@ -1,9 +1,32 @@
...
+    end
+
+    def dump_schema_task(as_regexp = false)
+      str =
+      if Rails.version < '7.0'
+        "db:structure:dump"
+      else
+        "db:schema:dump"
+      end
+      return as_regexp ? Regexp.new(str) : str
```
Comment (tagliala) [inline spec/support/aruba.rb:27]: useless return

File: Rakefile (hunk 16 above)
Comment (tagliala) [inline Rakefile:2]: let's avoid this require and use both options `--skip-webpack-install --skip-javascript`

File: Rakefile [hunk 20]
```diff
@@ -16,7 +17,11 @@ namespace :testapp do
     FileUtils.mkdir_p('tmp/aruba')
     Dir.chdir('tmp') do
       FileUtils.rm_rf('railsapp')
-      sh 'rails new railsapp --skip-bundle --skip-webpack-install --skip-git'
+      if Rails.version < '7.0'
+        sh 'rails new railsapp --skip-bundle --skip-webpack-install --skip-git'
+      else
+        sh 'rails new railsapp --skip-bundle --skip-javascript --skip-git'
+      end
```
Comment (tagliala) [inline Rakefile:24]: use both options and skip this conditional

File: Rakefile [hunk 21]
```diff
@@ -16,7 +16,7 @@ namespace :testapp do
     FileUtils.mkdir_p('tmp/aruba')
     Dir.chdir('tmp') do
       FileUtils.rm_rf('railsapp')
-      sh 'rails new railsapp --skip-bundle --skip-webpack-install --skip-git'
+        sh 'rails new railsapp --skip-bundle --skip-javascript --skip-webpack-install --skip-git'
```
Comment (tagliala) [inline Rakefile:19]: wrong indentation

**Review by tagliala — CHANGES_REQUESTED** [review]

File: lib/chrono_model/conversions.rb [hunk 22]
```diff
@@ -15,7 +15,7 @@ def string_to_utc_time(string)
     end
 
     def time_to_utc_string(time)
-      [time.to_s(:db), sprintf('%06d', time.usec)].join '.'
+      [time.to_formatted_s(:db), sprintf('%06d', time.usec)].join '.'
```
Comment (tagliala) [inline lib/chrono_model/conversions.rb:18]: Please split this change in a different commit, it will be useful for the rebase

I would like to reduce the number of changes and commits in this branch before merging

File: Rakefile [hunk 23]
```diff
@@ -17,11 +16,7 @@ namespace :testapp do
     FileUtils.mkdir_p('tmp/aruba')
     Dir.chdir('tmp') do
       FileUtils.rm_rf('railsapp')
-      if Rails.version < '7.0'
-        sh 'rails new railsapp --skip-bundle --skip-webpack-install --skip-git'
-      else
-        sh 'rails new railsapp --skip-bundle --skip-javascript --skip-git'
-      end
+        sh 'rails new railsapp --skip-bundle --skip-javascript --skip-webpack-install --skip-git'
     end
```
Comment (tagliala) [inline Rakefile:20]: please split changes to rake file in a different commit
